File: src/content/examples/math-interpreter-ts/src/evaluator.ts
import type { ASTNode } from "./types.js";

// The Evaluator walks through the AST and computes the result
// It uses recursion to evaluate nested expressions
export class Evaluator {
	// Recursively evaluate an AST node and return the numeric result
	public evaluate(node: ASTNode): number {
		// Base case: if it's a number, just return its value
		if (node.type === "Number") {
			return node.value;
		}

		// Recursive case: if it's an operation, evaluate both sides first
		if (node.type === "BinaryOp") {
			const left = this.evaluate(node.left);
			const right = this.evaluate(node.right);

			// Then apply the operator
			switch (node.operator) {
				case "+":
					return left + right;
				case "-":
					return left - right;
				case "*":
					return left * right;
				case "/":
					if (right === 0) {
						throw new Error("Division by zero");
					}
					return left / right;
			}
		}

		// This should never happen if our parser is correct
		throw new Error(`Unknown node type`);
	}
}

File: src/content/examples/math-interpreter-c/evaluator.c
#include "evaluator.h"

/*
 * Recursively evaluate an AST node
 * For numbers: return the value
 * For operations: evaluate children and apply the operator
 */
double evaluate(const ASTNode *node, const char **error) {
  if (!node) {
    *error = "Invalid AST node";
    return 0.0;
  }

  /* Base case: leaf node with a number */
  if (node->type == NODE_NUMBER) {
    return node->data.number;
  }

  /* Recursive case: binary operation */
  if (node->type == NODE_BINARY_OP) {
    /* Evaluate both operands recursively */
    double left = evaluate(node->data.binary_op.left, error);
    if (*error)
      return 0.0;

    double right = evaluate(node->data.binary_op.right, error);
    if (*error)
      return 0.0;

    /* Apply the operator */
    switch (node->data.binary_op.operator) {
    case TOKEN_PLUS:
      return left + right;
    case TOKEN_MINUS:
      return left - right;
    case TOKEN_MULTIPLY:
      return left * right;
    case TOKEN_DIVIDE:
      if (right == 0.0) {
        *error = "Division by zero";
        return 0.0;
      }
      return left / right;
    default:
      *error = "Invalid operator";
      return 0.0;
    }
  }

  *error = "Unknown node type";
  return 0.0;
}
